### ISSUE TITLE
vagrant-wrapper issues with `kitchen test`

### ISSUE THREAD

**Issue by tknerr** [test-kitchen/test-kitchen#575]
Running `kitchen test` from within [this bundle](https://github.com/tknerr/sample-toplevel-cookbook/blob/915f3b6d83d97d3b25c74c7fc74c850ea9458bfb/Gemfile) causes this error:
```
tkn@dev-box:~/sample-toplevel-cookbook$ bundle exec kitchen test
-----> Starting Kitchen (v1.2.1)
-----> Cleaning up any prior instances of <default-ubuntu-1204-i386>
-----> Destroying <default-ubuntu-1204-i386>...
       Finished destroying <default-ubuntu-1204-i386> (0m0.00s).
-----> Testing <default-ubuntu-1204-i386>
-----> Creating <default-ubuntu-1204-i386>...
       Bringing machine 'default' up with 'lxc' provider...
       The following berks command failed to execute:
       
           /opt/chefdk/embedded/lib/ruby/gems/2.1.0/bin/berks --version --format json
       
       The stdout and stderr are shown below:
       
           stdout: 
           stderr: /opt/vagrant/embedded/lib/ruby/2.0.0/rubygems/dependency.rb:296:in `to_specs': Could not find 'berkshelf' (>= 0) among 59 total gem(s) (Gem::LoadError)
       	from /opt/vagrant/embedded/lib/ruby/2.0.0/rubygems/dependency.rb:307:in `to_spec'
       	from /opt/vagrant/embedded/lib/ruby/2.0.0/rubygems/core_ext/kernel_gem.rb:47:in `gem'
       	from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/bin/berks:22:in `<main>'
>>>>>> ------Exception-------
>>>>>> Class: Kitchen::ActionFailed
>>>>>> Message: Failed to complete #create action: [Expected process to exit with [0], but received '1'
---- Begin output of vagrant up --no-provision --provider=lxc ----
STDOUT: Bringing machine 'default' up with 'lxc' provider...
STDERR: The following berks command failed to execute:

    /opt/chefdk/embedded/lib/ruby/gems/2.1.0/bin/berks --version --format json

The stdout and stderr are shown below:

    stdout: 
    stderr: /opt/vagrant/embedded/lib/ruby/2.0.0/rubygems/dependency.rb:296:in `to_specs': Could not find 'berkshelf' (>= 0) among 59 total gem(s) (Gem::LoadError)
	from /opt/vagrant/embedded/lib/ruby/2.0.0/rubygems/dependency.rb:307:in `to_spec'
	from /opt/vagrant/embedded/lib/ruby/2.0.0/rubygems/core_ext/kernel_gem.rb:47:in `gem'
	from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/bin/berks:22:in `<main>'
---- End output of vagrant up --no-provision --provider=lxc ----
Ran vagrant up --no-provision --provider=lxc returned 1]
>>>>>> ----------------------
>>>>>> Please see .kitchen/logs/kitchen.log for more details
>>>>>> Also try running `kitchen diagnose --all` for configuration
```

Yep, I found these two related issues:
* https://github.com/berkshelf/berkshelf/issues/1357
* https://github.com/opscode/chef-dk/issues/282

However, what's weird is this:
```
tkn@dev-box:~/sample-toplevel-cookbook$ which berks
/opt/chefdk/bin/berks
tkn@dev-box:~/sample-toplevel-cookbook$ berks --version --format json
{
  "version": "3.2.1"
}
tkn@dev-box:~/sample-toplevel-cookbook$ bundle exec which berks
/home/tkn/.chefdk/gem/ruby/2.1.0/bin/berks
tkn@dev-box:~/sample-toplevel-cookbook$ bundle exec berks --version --format json
{
  "version": "3.2.2"
}
@dev-box:~/sample-toplevel-cookbook$ /opt/chefdk/bin/berks --version --format json
{
  "version": "3.2.1"
}
tkn@dev-box:~/sample-toplevel-cookbook$ /opt/chefdk/embedded/lib/ruby/gems/2.1.0/bin/berks --version --format json
/opt/chefdk/embedded/lib/ruby/site_ruby/2.1.0/rubygems/specification.rb:2087:in `check_version_conflict': can't activate json-1.7.7, already activated json-1.8.1 (Gem::LoadError)
	from /opt/chefdk/embedded/lib/ruby/site_ruby/2.1.0/rubygems/specification.rb:1278:in `activate'
	from /opt/chefdk/embedded/lib/ruby/site_ruby/2.1.0/rubygems/core_ext/kernel_require.rb:90:in `block in require'
	from /opt/chefdk/embedded/lib/ruby/site_ruby/2.1.0/rubygems/core_ext/kernel_require.rb:89:in `each'
	from /opt/chefdk/embedded/lib/ruby/site_ruby/2.1.0/rubygems/core_ext/kernel_require.rb:89:in `require'
	from /home/tkn/.chefdk/gem/ruby/2.1.0/gems/berkshelf-3.2.2/lib/berkshelf.rb:5:in `<top (required)>'
	from /opt/chefdk/embedded/lib/ruby/site_ruby/2.1.0/rubygems/core_ext/kernel_require.rb:69:in `require'
	from /opt/chefdk/embedded/lib/ruby/site_ruby/2.1.0/rubygems/core_ext/kernel_require.rb:69:in `require'
	from /home/tkn/.chefdk/gem/ruby/2.1.0/gems/berkshelf-3.2.2/lib/berkshelf/cli.rb:1:in `<top (required)>'
	from /opt/chefdk/embedded/lib/ruby/site_ruby/2.1.0/rubygems/core_ext/kernel_require.rb:69:in `require'
	from /opt/chefdk/embedded/lib/ruby/site_ruby/2.1.0/rubygems/core_ext/kernel_require.rb:69:in `require'
	from /home/tkn/.chefdk/gem/ruby/2.1.0/gems/berkshelf-3.2.2/bin/berks:3:in `<top (required)>'
	from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/bin/berks:23:in `load'
	from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/bin/berks:23:in `<main>'
```

Why does test-kitchen call out to the `/opt/chefdk/embedded/lib/ruby/gems/2.1.0/bin/berks` executable rather than using `/opt/chefdk/bin/berks` or better `/home/tkn/.chefdk/gem/ruby/2.1.0/bin/berks` when running in a bundle context?

**Comment by tknerr** [comment]
@danielsdeleo FYI it works on my [windows based setup](https://github.com/tknerr/bills-kitchen), but still fails on my [ubuntu dev-box](https://github.com/tknerr/dev-box/) (both with ChefDK 0.3.5).

These are the env vars on windows:
```
W:\repo\sample-toplevel-cookbook>set | grep BUNDLE

W:\repo\sample-toplevel-cookbook>set | grep RUBY

W:\repo\sample-toplevel-cookbook>set | grep GEM
```

...and with the bundle context:
```
W:\repo\sample-toplevel-cookbook>bundle exec set | grep BUNDLE
BUNDLE_BIN_PATH=W:/tools/chefdk/embedded/lib/ruby/gems/2.0.0/gems/bundler-1.7.5/bin/bundle
BUNDLE_GEMFILE=W:/repo/sample-toplevel-cookbook/Gemfile

W:\repo\sample-toplevel-cookbook>bundle exec set | grep RUBY
RUBYLIB=W:/tools/chefdk/embedded/lib/ruby/gems/2.0.0/gems/bundler-1.7.5/lib
RUBYOPT=-rbundler/setup

W:\repo\sample-toplevel-cookbook>bundle exec set | grep GEM
BUNDLE_GEMFILE=W:/repo/sample-toplevel-cookbook/Gemfile
GEM_HOME=W:/tools/chefdk/embedded/lib/ruby/gems/2.0.0
GEM_PATH=W:/tools/chefdk/embedded/lib/ruby/gems/2.0.0;W:/home/.chefdk/gem/ruby/2.0.0
```

One notable difference is that on the windows setup the `GEM_PATH` looks into the omnibus ruby gem dir first, then in the `~/.chefdk` gem dir next, while on ubuntu it looks into the user home gem dir first.

I remember having some problems with the user home gem installs on windows (might have been exactly a bundler problem like this, but don't remember the details), thus I [explicitly disabled the user home gem installs](https://github.com/tknerr/bills-kitchen/commit/5c8b04b2e797a669e2e4dc79a019a0ab217be80a).

**Comment by tknerr** [comment]
FYI - I just re-enabled the user home gem installs and it still works on my windows setup. So that seems not to be the culprit.

**Comment by tknerr** [comment]
Finally I was able to track down some further details concerning the bundler issue:

 * first of all, this only occurs if the `vagrant-wrapper` gem is part of the bundle. After [removing it from the bundle](https://github.com/tknerr/sample-toplevel-cookbook/commit/e9105869694ed3bdf14943e1b7d3ce5927302fc0) it worked as expected
   * if it's not part of the bundle, bundler will correctly look under `~/.chefdk/gem/ruby/2.1.0`
   * if it is part of the bundle, bundler will look under `~/.bundle/gem/ruby/2.1.0` which is empty

I have no clue how org-binbab/gem-vagrant-wrapper could break this, as I see it nowhere mangling with bundler or bundler paths. It [does an exec](https://github.com/org-binbab/gem-vagrant-wrapper/blob/master/lib/vagrant-wrapper.rb#L172) however, which could have something to do with it.

As a workaround I now created a `~/.bundle/config` which explicitly sets `BUNDLE_PATH` to `~/.chefdk/gem/ruby/2.1.0`. This is being preserved during the test-kitchen run and makes it work again in both cases, i.e. with vagrant-wrapper in the Gemfile and without.

@danielsdeleo would it be worthwile (and correct!) exporting `BUNDLE_PATH` via `chef shell-init`?

/cc @binarybabel @btm
